File: catalog/backends/java-spring/patterns/CustomMetricsService.java
package com.example.service;

import com.timgroup.statsd.NonBlockingStatsDClient;
import com.timgroup.statsd.StatsDClient;
import org.springframework.web.bind.annotation.*;
import java.util.Map;

// Pattern: DogStatsD — custom gauge and counter
// Adapt: metric names should reflect domain (e.g., orders.pending)
@RestController
public class CustomMetricsService {

    private final StatsDClient statsd = new NonBlockingStatsDClient(
        "", "datadog-agent", 8125
    );

    @PostMapping("/api/metrics/record")
    public Map<String, String> record(@RequestParam String name, @RequestParam double value) {
        statsd.gauge(name, value);
        statsd.incrementCounter(name + ".calls");
        return Map.of("status", "recorded", "metric", name);
    }
}

File: catalog/backends/python-flask/patterns/custom_metrics_service.py
from datadog import statsd
from flask import request, jsonify

# Pattern: DogStatsD — custom gauge and counter
# Adapt: metric names should reflect domain (e.g., orders.pending)


def record_metric():
    name = request.args.get("name", "custom.metric")
    value = float(request.args.get("value", "1"))
    statsd.gauge(name, value)
    statsd.increment(f"{name}.calls")
    return jsonify({"status": "recorded", "metric": name})
